torch/_functorch/autograd_function.py

```diff
@@ -423,29 +420,31 @@ def inner(inputs, outputs):
             # to the correct shape.
             # See NOTE: [Why can't we rely on autograd to reduce expanded gradients?]
             # for more details
-            nonlocal input_shapes_
             input_shapes_ = tuple(
                 inp.shape if isinstance(inp, torch.Tensor) else None for inp in inputs
             )
-            nonlocal saved_tensors_bdims_
-            saved_tensors_bdims_ = wrapped_ctx._pt_saved_tensors_bdims
+            if not hasattr(ctx, "input_shapes"):
+                ctx.input_shapes = {}
+            ctx.input_shapes.update({key: input_shapes_})
```

Comment:
We should try to avoid collisions (like if the original autograd.Function had an input_shapes field that was not added by us). So let's name our preface our fields  with "`_pt_`" (that's how we did it previously with e.g. `_pt_saved_tensros_bdims`)